backend/src/github-util.ts

```diff
@@ -0,0 +1,242 @@
+      created_at: Date.parse(pr.created_at),
+      diff_size: pr.additions + pr.deletions
+    };
+  });
+};
+
+/** Creates 'valid' result if no errors are raised, 'invalid' result otherwise.  */
+const createValidationResult = async (validation_function): Promise<ValidationResult> => {
+  try {
+    await validation_function();
+  } catch (err) {
+    if (err instanceof Error) {
+      return {
+        status: 'invalid',
+        reason: err.message
+      };
+    }
+  }
+
+  // if no errors encountered, then valid submission
+  return { status: 'valid' };
+};
+
+/** Determines whether PR review is valid. */
+const validateReview = async (
+  portfolio: DevPortfolio,
+  submission: DevPortfolioSubmission,
+  review_url: string
+): Promise<ValidationResult> => {
+  const start = portfolio.earliestValidDate;
+  const end = portfolio.deadline;
+  const username = submission.member.github || ''; // must be github username
+
+  return createValidationResult(async () => {
+    // get review object
+    const review = await getReviewedPR(parseGithubUrl(review_url));
+
+    // cannot review own PR
+    if (review.created_by === username) {
+      throw new Error(`Cannot use PR ${review.url} opened by user for review requirement.`);
+    }
+
+    // comments made by user within the date range
+    const eligible_comments = filterComments(review.comments, username, start, end);
+
+    // placeholder logic: valid if at least 10 words total
+    const total_word_count = eligible_comments.reduce(
+      (count, comment) => count + comment.content.split(' ').length,
+      0
+    );
+    if (total_word_count < 10) {
+      throw new Error('Trivial review.');
+    }
+  });
+};
+
+/** Determines whether an open PR is valid. */
+const validateOpen = async (
+  portfolio: DevPortfolio,
+  submission: DevPortfolioSubmission,
+  open_url: string
+): Promise<ValidationResult> => {
+  const start = portfolio.earliestValidDate;
+  const end = portfolio.deadline;
+  const username = submission.member.github || ''; // must be github username
+
+  return createValidationResult(async () => {
+    // get open object
+    const open = await getOpenedPR(parseGithubUrl(open_url));
+
+    if (open.created_by !== username) {
+      throw new Error(`User ${username} did not open the pull request ${open.url}.`);
+    }
+
+    if (start >= open.created_at && open.created_at >= end) {
+      const start_date = new Date(start).toDateString();
+      const end_date = new Date(end).toDateString();
+      throw new Error(
+        `Pull request ${open.url} was not made between ${start_date} and ${end_date}.`
+      );
+    }
+
+    // placeholder logic: valid if at least 10 changes total
+    if (open.diff_size < 10) {
+      throw new Error('Trivial PR.');
+    }
+  });
+};
+
+/** ="at least one of `results` is valid" */
+const atLeastOneValid = (results: ValidationResult[]) =>
+  results.reduce((isValid, result) => isValid || result.status === 'valid', false);
```

Comment:
[`some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) might be more suitable here

Reply:
Didn't know that existed!